File: pkg/twrfiles/format_air.go
package twrfiles

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// FormatAIR serializes aircraft rows to TWRTrainer-compatible .air text.
//
// Normative contract (docs/design/apt-air-editor.md):
//   - Exactly 16 colon-separated fields per row; no trailing colon
//   - Callsign, Type, Engine, Rules, Dep, Arr, XPDRMode uppercased
//   - Lat/Lon always %.6f; Alt/Speed/Heading compact (integer form when whole)
//   - Route/Remarks as stored; LF between lines; trailing newline when non-empty
//   - Slice order preserved; no field-legend comment block
func FormatAIR(rows []Aircraft) string {
	if len(rows) == 0 {
		return ""
	}
	var b strings.Builder
	for i, ac := range rows {
		if i > 0 {
			b.WriteByte('\n')
		}
		writeAircraftLine(&b, ac)
	}
	b.WriteByte('\n')
	return b.String()
}

func writeAircraftLine(b *strings.Builder, ac Aircraft) {
	// 0 callsign
	b.WriteString(strings.ToUpper(ac.Callsign))
	b.WriteByte(':')
	// 1 type
	b.WriteString(strings.ToUpper(ac.Type))
	b.WriteByte(':')
	// 2 engine
	b.WriteString(strings.ToUpper(ac.Engine))
	b.WriteByte(':')
	// 3 rules
	b.WriteString(strings.ToUpper(ac.Rules))
	b.WriteByte(':')
	// 4 dep
	b.WriteString(strings.ToUpper(ac.Dep))
	b.WriteByte(':')
	// 5 arr
	b.WriteString(strings.ToUpper(ac.Arr))
	b.WriteByte(':')
	// 6 cruise alt (integer)
	b.WriteString(strconv.Itoa(ac.CruiseAlt))
	b.WriteByte(':')
	// 7 route (as stored)
	b.WriteString(ac.Route)
	b.WriteByte(':')
	// 8 remarks (as stored)
	b.WriteString(ac.Remarks)
	b.WriteByte(':')
	// 9 squawk
	b.WriteString(ac.Squawk)
	b.WriteByte(':')
	// 10 xpdr mode
	b.WriteString(strings.ToUpper(ac.XPDRMode))
	b.WriteByte(':')
	// 11 lat
	b.WriteString(fmt.Sprintf("%.6f", ac.Lat))
	b.WriteByte(':')
	// 12 lon
	b.WriteString(fmt.Sprintf("%.6f", ac.Lon))
	b.WriteByte(':')
	// 13 alt
	b.WriteString(formatAirCompactFloat(ac.Alt))
	b.WriteByte(':')
	// 14 speed
	b.WriteString(formatAirCompactFloat(ac.Speed))
	b.WriteByte(':')
	// 15 heading
	b.WriteString(formatAirCompactFloat(ac.Heading))
}

// formatAirCompactFloat emits an integer decimal string when v is a whole
// number; otherwise the shortest ParseFloat-round-tripping form.
func formatAirCompactFloat(v float64) string {
	if !math.IsNaN(v) && !math.IsInf(v, 0) && v == math.Trunc(v) {
		return strconv.FormatInt(int64(v), 10)
	}
	return strconv.FormatFloat(v, 'f', -1, 64)
}

File: internal/web/static/js/openfsd/airport-editor/format-air.js
/**
 * Format aircraft rows → TWRTrainer-compatible .air text.
 * Port of pkg/twrfiles.FormatAIR — normative Format contract (16 fields).
 */

/**
 * Serialize aircraft rows. Empty list → empty string (no trailing newline).
 * @param {import('./model.js').Aircraft[]} rows
 * @returns {string}
 */
export function formatAIR(rows) {
  if (!Array.isArray(rows) || rows.length === 0) return '';
  const lines = rows.map(writeAircraftLine);
  return `${lines.join('\n')}\n`;
}

/**
 * @param {import('./model.js').Aircraft} ac
 * @returns {string}
 */
function writeAircraftLine(ac) {
  const fields = [
    String(ac.callsign ?? '').toUpperCase(),
    String(ac.type ?? '').toUpperCase(),
    String(ac.engine ?? '').toUpperCase(),
    String(ac.rules ?? '').toUpperCase(),
    String(ac.dep ?? '').toUpperCase(),
    String(ac.arr ?? '').toUpperCase(),
    String(Math.trunc(Number(ac.cruiseAlt) || 0)),
    ac.route ?? '',
    ac.remarks ?? '',
    ac.squawk ?? '',
    String(ac.xpdrMode ?? '').toUpperCase(),
    formatCoord6(ac.lat),
    formatCoord6(ac.lon),
    formatAirCompactFloat(ac.alt),
    formatAirCompactFloat(ac.speed),
    formatAirCompactFloat(ac.heading),
  ];
  return fields.join(':');
}

/**
 * Always six fractional digits (Go %.6f).
 * @param {number} v
 */
function formatCoord6(v) {
  return Number(v).toFixed(6);
}

/**
 * Integer form when whole; else shortest decimal (Go FormatFloat 'f' -1).
 * @param {number} v
 */
export function formatAirCompactFloat(v) {
  const n = Number(v);
  if (!Number.isFinite(n)) return '0';
  if (n === Math.trunc(n)) {
    return String(Math.trunc(n));
  }
  let s = String(n);
  if (/e/i.test(s)) {
    s = n.toFixed(16).replace(/(\.\d*?)0+$/, '$1').replace(/\.$/, '');
  }
  return s;
}
